laplace/curvature/asdfghjkl.py

```diff
@@ -101,36 +113,47 @@ def _get_kron_factors(self, curv, M):
         return Kron(kfacs)
 
     @staticmethod
-    def _rescale_kron_factors(kron, N):
+    def _rescale_kron_factors(kron: Kron, N: int) -> Kron:
         for F in kron.kfacs:
             if len(F) == 2:
                 F[1] *= 1 / N
         return kron
 
-    def diag(self, X, y, **kwargs):
+    def diag(
+        self,
+        x: torch.Tensor | MutableMapping[str, torch.Tensor | Any],
+        y: torch.Tensor,
+        **kwargs: dict[str, Any],
+    ) -> tuple[torch.Tensor, torch.Tensor]:
         with torch.no_grad():
             if self.last_layer:
-                f, X = self.model.forward_with_features(X)
+                f, x = self.model.forward_with_features(x)
             else:
-                f = self.model(X)
+                f = self.model(x)
             loss = self.lossfunc(f, y)
         curv = fisher_for_cross_entropy(
-            self._model, self._ggn_type, SHAPE_DIAG, inputs=X, targets=y
+            self._model, self._ggn_type, SHAPE_DIAG, inputs=x, targets=y
         )
         diag_ggn = curv.matrices_to_vector(None)
         if self.subnetwork_indices is not None:
             diag_ggn = diag_ggn[self.subnetwork_indices]
         return self.factor * loss, self.factor * diag_ggn
 
-    def kron(self, X, y, N, **kwargs):
+    def kron(
+        self,
+        x: torch.Tensor | MutableMapping[str, torch.Tensor | Any],
+        y: torch.Tensor,
+        N: int,
+        **kwargs: dict[str, Any],
+    ) -> tuple[torch.Tensor, Kron]:
         with torch.no_grad():
             if self.last_layer:
-                f, X = self.model.forward_with_features(X)
+                f, x = self.model.forward_with_features(x)
             else:
-                f = self.model(X)
+                f = self.model(x)
             loss = self.lossfunc(f, y)
         curv = fisher_for_cross_entropy(
```

Comment:
The only thing that's still missing is the type for `curv` here. @aleximmer @runame do you know what's its type? I dug into ASDL v0.1's code and came back even more confused...

Reply:
This function does not exist in ASDL, only in its predecessor `asdfghjkl`. The return type will be `FisherExactCrossEntropy`, `FisherMCCrossEntropy`, or `FisherEmpCrossEntropy` depending on `self._ggn_type`. You should be able to either import these classes from `asdfghjkl` directly or from `asdfghjkl.fisher`.